man/htick.1
.TH HTICK 1 "Husky Tick v1.9.0" "2020-11-04" "Husky - Portable Fido Software"
.SH NAME
htick \- Husky File-Ticker
.SH SYNOPSIS
.B htick [-h] [-v] [-q] [-c config-file] [<command>...]
.SH "DESCRIPTION"
.PP
Htick is a part of Husky Fidonet software package. It handles thematic subscriptions to
files. Each such subscription forms a file area also known as a file echo. Every file
sent from a file echo to a subscribed sysop is accompanied with a ticket file with the
filename extension .tic containing technical information which allows correct handling
of the received file.
.SH "OPTIONS"
.B -h
- Print program name, version and usage information to screen and exit.
.TP
.B -v
- Print program name and version to screen and exit.
.TP
.B -q
- Quiet mode: don't print any messages to screen
.TP
.B -c config-file
- Specify an alternative config file
.SS "COMMANDS"
You may place several commands one after another with one exclusion: htick will
not execute any other command after \fBhatch\fP.
.IP \fBtoss\fP
Read a *.tic file and move the file to which it refers to the directory of the
file echo specified in the .tic file. After that the .tic is deleted.
.IP
If some errors preventing tossing are found during parsing the .tic, the .tic is
renamed and tossing is not executed. If the .tic contains wrong password, it is
renamed to .sec. If there is some error during reading the .tic, it is renamed
to .acs. If the .tic has wrong format i.e. does not conform to specification,
it is renamed to .bad. And if the .tic is not to us, it is renamed to .ntu.
.IP "\fBtoss -b\fP"
Try to toss ticket files with file extensions .sec, .acs, .bad and .ntu.
It is assumed that you have already fixed the errors that led to the renaming
of the files.
.IP "\fBscan\fP"
Scan the \fIRobotsArea\fR for mails to filefix, or first netmail area if \fIRobotsArea\fR
is not defined. If at least one such mail is found, the filefix commands
contained in the mail are executed. See also the 
\(dq\&FILEFIX COMMANDS\(dq\& section.
.IP "\fBffix [-f] [-s] <address> <filefix command>\fP"
Send the \fB<filefix command>\fP to our filefix from the name of \fB<address>\fP.
Our filefix will send a reply to the \fB<filefix command>\fP to the \fB<address>\fP.
All filefix commands are listed below in the \(dq\&FILEFIX COMMANDS\(dq\& section.
.RS 7
.IP \fB-f\fP
- also send the same command to the link's filefix
.IP \fB-s\fP
- do not send reply from our filefix to the link
.RE
.IP "\fBffix! <address> <filefix command>\fP"
The same as \fBffix -f <address> <command>\fP
.IP "\fBrelink <pattern> <address>\fP"
Refresh subscription for the areas matching the pattern. Htick sends to the
filefix at the \fB<address>\fP a list of subscription commands for all file
areas matching the \fB<pattern>\fP for which we are already subscribed and
which are not paused.
.IP "\fBresubscribe <pattern> <from-address> <to-address>\fP"
Move subscription for the areas matching the pattern from one link to another.
.IP "\fBresubscribe -f <file> <from-address> <to-address>\fP"
Move subscription from one link to another of the areas matching the area
patterns listed in this file with one pattern on a line.
.IP "\fBclean\fP"
Clean passthrough dir and old files in fileechos.
.IP
Every .tic file refers to some filename of a file we have received from a file
echo. "htick \fBclean\fP" collects all such filenames to which .tic files in the directories
defined by \fIbusyFileDir\fR and \fIticOutbound\fR refer and then checks every not .tic
file in \fIpassFileAreaDir\fR if it is a member of the collection. All files not
refered by any .tic are deleted.
.IP
A file echo declaration may have \fI-p <number of days>\fR option, which says
that the files in the file echo older than the <number of days> should be
purged. So the \fBclean\fP command deletes such files.
.IP
If \fISaveTic\fR statements are defined and contain \fI<DaysToKeepTics>\fR, the .tic
files older than \fI<DaysToKeepTics>\fR are purged.
.IP "\fBannounce\fP"
Announce new files. For announcing to work you have to define \fIAnnounceSpool\fR in
htick config. Announces are made to the echo defined in \fIReportTo\fR. If \fIReportTo\fR
is not defined, then announces are sent to the first netmail area.
.IP "\fBannfecho <file>\fP"
Announce new file echos in the text file so you may process the file with a script.
.IP "\fBsend <file> <filearea> <address>\fP"
Send the file from the filearea to the address. It will be sent together with
a freshly generated .tic so for the system at the <address> it will look as a
regular file from the file echo.
.IP "\fBhatch <file> <area> [replace [<filemask>]] [desc <desc> [<ldesc>]]\fP"
Send the file to the file area together with a generated ticket file (with .tic
filename extension). The \fBhatch\fP command should be the last one if several
commands are used in one htick run, that is no other command is executed after
the \fBhatch\fP.
.IP
.RS 7
.IP \fB<file>\fP
- the file to send (the parameter is mandatory);
.IP \fB<area>\fP
- the file area to which we want to send the file (the parameter is
mandatory). All the rest parameters are optional;
.IP "\fBreplace [<filemask>]\fP"
- this option is used to replace file(s) sent to the same file echo earlier.
\fB<filemask>\fP is the mask of the files that at subscriber systems will be
replaced by the file we send now. If the \fB<filemask>\fP is omitted, then the
filename coinsiding with the filename we send will be replaced at subscriber
systems;
.IP "\fBdesc <desc> [<ldesc>]\fP"
- this option specifies the description of the sent file, \fB<desc>\fP is a short
one-line description, and \fB<ldesc>\fP is a multi-line description;
.IP
You may use a string in single or double quotes as \fB<desc>\fP. You may also use the
following macros:
.RS 7
.IP "\fB@@BBS\fP"
- to substitute the first line from the \fIfiles.bbs\fR as \fB<desc>\fP;
.IP "\fB@@DIZ\fP"
- to substitute the first line from the \fIfile_id.diz\fR as \fB<desc>\fP;
.IP "\fB@@<file>\fP"
- to substitute the first line from the \fB<file>\fP as \fB<desc>\fP;
.RE
.IP
You may use the following macros as \fB<ldesc>\fP:
.RS 7
.IP "\fB@BBS\fP"
- to substitute the whole \fIfiles.bbs\fR file as \fB<ldesc>\fP;
.IP "\fB@DIZ\fP"
- to substitute the whole \fIfile_id.diz\fR file as \fB<ldesc>\fP;
.IP "\fB@<file>\fP"
- to substitute the whole \fB<file>\fP as \fB<ldesc>\fP.
.RE
There are some recommendations for hatching files to a file echo. It is better
to add a file with the name \fIfile_id.diz\fR containing a short description of
what you send. The lines in the file should not be longer than 45 characters
and the number of lines is preferably no more than 10. Please put the file(s)
you send together with \fIfile_id.diz\fR in a zip archive. The archive filename
MUST be in 8.3 format. Good manners suggest that you use \fBdesc <desk>\fP
at sending. You may also add \fB<ldesc>\fP if you wish. If you want to resend
the archive you have already sent earlier, please do not forget to use \fBreplace [<filemask>]\fP.
.RE
.IP "\fBfilelist <file> [<dirlist>]\fP"
Generate a file list and write it to the \fB<file>\fR. The file list includes
all files of all not hidden file echos and files of your BBS. A file echo is hidden
if its declaration contains \fB-hide\fP option. For every file echo or BBS file
area the file list contains 3 items. The first one is file area name and
description. The second one is a list of files in which for every file the
following is specified: filename, file size, file date, file description. The
third item is total number of files in the file area and their total size. And
at the end we see the total number of files in all file areas and their total
size. \fB<dirlist>\fR is a file to which a directory list of all file areas is
written.
.RE
.SH "FILEFIX COMMANDS"
Filefix is a robot supported by htick that allows handling subscription to
file areas at an uplink system for a node or at the boss node for a point.
In order to send a command to the filefix, the sysop writes a message to it,
specifying the filefix password in the subject line, and the filefix commands
in the body of the message, one command per line. One should put \fBFilefix\fP
as an addressee name and the uplink (boss node) Fidonet address as its address.
The \fBFilefix\fP name is case insensitive. Filefix commands received by our
system from downlinks and points are handled by htick when it runs \fBhtick scan\fP.
.PP
Another option of using filefix commands is \fBhtick ffix <address> <filefix command>\fP.
It allows us to run a filefix command from the name of the system with Fidonet address
\fB<address>\fP. In this case, the sysop with the address \fB<address>\fP will receive
the filefix response.

.IP "\fB%list\fP"
- send me the list of available file areas.
.IP "\fB%help\fP"
- send me the list of filefix commands.
.IP "\fB%unlinked\fP"
- send me the list of not linked file areas.
.IP "\fB%linked\fP"
- send me the list of linked file areas.
.IP "\fB%query\fP"
- the same as \fB%linked\fP
.IP "\fB%avail\fP"
- send me the list of file areas available at your uplinks.
.IP "\fB%pause\fP"
- temporarily stop sending me files from file echos. Information
about the subscription is not lost.
.IP "\fB%resume\fP"

HISTORY
HTICK 1.4-stable
Changes between htick-1.0-stable and htick-1.4-stable

1. MAJOR feature improvements:
   - configurable announces about received files;
   - purging for file echos with deleting description from files.bbs;
   - New command line option: "-c config-file" - specify alternate config
     file (optional);
   - Change statistics information in log file;
   - Allow unconditional forward subscribe when forwardFileRequestsFile for
     link not defined in config;
   - New command 'ffix!' in addition to 'ffix'. It does the same job as 'ffix'
     and send subscride/unsubscribe letter to link that was manualy
     subscribed/unsubscribed by filefix command	like:
        htick ffix! 2:464/910 -910.TEST
     This command remove file echo link 2:464/910 from 910.TEST and produce
     letter to 2:464/910 with body "-910.TEST";
   - Write documentation at texinfo format;
   - Return non-zero exit code if hatching unsussessful.

2. CONFIG changes:
   - area definition options -manual & -mandatory are different:
     '-mandatory' prevent unsubscribe, '-manual' disable subscribe;
   - FileArea parameter '-p <number>' : purge fileecho (files age in days);
   - AnnounceSpool : directory for temporary announce files;
   - AnnAreaTag : Name of area when annouce will be placed, begins new
     announce definition section;
   - AnnAddrFrom : announce message sender address;
   - AnnFrom : announce message sender name;
   - AnnAddrTo : announce message recipient address;
   - AnnTo : announce message recipient name;
   - AnnSubj : announce message subject;
   - AnnInclude : Array of fileecho names|masks that will be announced in
     this annAreaTag;
   - AnnExclude : Array of fileecho names|masks that won't be announced
     in this annAreaTag;
   - AnnMessFlags : announce message flags ("k/s", "loc", "hld", ...);
   - AnnOrigin : origin for announce messages;
   - AnnFileOrigin : show system address that hatch file in announce;
   - AnnFileRFrom : show system address which file recived from;
   - AutoFileCreateFlag : create flag when filearea has been autocreated;
   - FileFixNames : set of names on which filefix will respond;
   - FileFixEchoLimit: limit fileechoes subscribe number for link;
   - delNotRecievedTIC renamed to delNotReceivedTIC (spelling fix)

3. IMPORTANT bugfixes:
   - fix forward subscribe;
   - fix CR-LF for autocreated "echoarea" lines;
   - fixed bug with missed files in report;
   - recode filearea descriptions;
   - save modification time for files in fileechoes;
   - set 'Loc' flag on messages from FileFix.
   - fixed crash while clean on passthrough areas with -p defined
   - replace char 026 in ldesc to space. some programs think that 026 is EOF.
   - save tics when hatch files



HTICK 1.0-stable (started 25.06.2002)

This file describe changes between htick-0.9 and htick-1.0

1. MAJOR feature improvements:
   - amiga style outbound suppurt;
   - cygwin port;
   - build htick/win32 with dlls (MS Visual C/C++);
   - new command line option: -q (quiet mode);
   - new log messages: 0-9A-D - user-level, E-Z - debug-level;
   - do filefix	before toss;
   - fileboxes support;
   - filefix from command line
   - use hardlinks if posible (unixes, NT);
   - Debuan Linux package;
   - man pages install & uninstall;
   - statistics generator (perl script);
   - forward requests to uplink;
   - '%avail' filefix command;
   - pause fileechoes only;
   - filelist building.

2. CONFIG changes:
   - FileFixFSC87Subset : ignore unknown keywords in tic-file
   - FileForwardRequest :
   - RemoteFileRobotName (in link definition) : Name of remote file (tic) robot
     (need for FileForwardRequest);
   - FileDescName : file with archive description (usually file_id.diz);
   - FileAreaDefaults : default parameters for autocreate filearea;
   - BBSArea : BBS files directory (for file list building);
   - FileEchoFlavour (in link definition) : set outbound flavor for fileecho
     files (Imm, Crash, Direct, Hold, Normal);
   - filearea option '-nodiz' : do not try to get description from
     <fileDescName>;
   - syntax of token 'SaveTic' expanded;
   - new token for link  "LinkFileFaseDir" : Where autocreated file echos
     will be placed (may be "passthrough");
   - pause statement syntax expanded;

3. IMPORTANT bugfixes:
   - don't toss .tic with HIDDEN attribute set (non-unix version)
   - fixed inbound file searching according to fsc
   - fix memory fault if charset tables not defined in config;
   - fix problem with autocreating;
   - fix hatching from fileechodir: don't delete file;
   - areadesc set into TIC at 'htick hatch';

src/htick.c
/*****************************************************************************
 * HTICK --- FTN Ticker / Request Processor
 *****************************************************************************
 * Copyright (C) 1999 by
 *
 * Gabriel Plutzar
 *
 * Fido:     2:31/1
 * Internet: gabriel@hit.priv.at
 *
 * Vienna, Austria, Europe
 *
 * This file is part of HTICK, which is based on HPT by Matthias Tichy,
 * 2:2432/605.14 2:2433/1245, mtt@tichy.de
 *
 * HTICK is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2, or (at your option) any
 * later version.
 *
 * HTICK is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with HTICK; see the file COPYING.  If not, write to the Free
 * Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *****************************************************************************
 * $Id$
 *****************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <sys/types.h>
#include <signal.h>
#include <fcntl.h>
#include <sys/stat.h>

#if (defined(__EMX__) || defined(__MINGW32__)) && defined(__NT__)
/* we can't include windows.h for prevent conflicts: send() & other */
# define CharToOem CharToOemA
#endif

/*  compiler.h */
#include <huskylib/compiler.h>

#ifdef HAS_UNISTD_H
# include <unistd.h>
#endif

#ifdef HAS_IO_H
# include <io.h>
#endif

#ifdef __OS2__
# define INCL_DOSPROCESS
# define INCL_DOSERRORS
# include <os2.h>
#endif

/* smapi */
#include <smapi/msgapi.h>
#include <huskylib/huskylib.h>

/* fidoconf */
#include <fidoconf/fidoconf.h>
#include <fidoconf/common.h>
#include <huskylib/log.h>
#include <huskylib/recode.h>
#include <huskylib/xstr.h>
#include <huskylib/ftnaddr.h>

/* areafix library */
#include <areafix/areafix.h>
#include <areafix/afglobal.h>

/* htick */
#include <htick.h>
#include <global.h>
#include "../cvsdate.h"
#include "version.h"
#include <toss.h>
#include <scan.h>
#include <hatch.h>
#include <filelist.h>
#include <htickafix.h>
#include "report.h"

int processHatchParams( int i, int argc, char **argv )
{

  char *basename, *extdelim;

  if( argc - i < 2 )
  {
    fprintf( stderr, "Insufficient number of arguments\n" );
    return 0;
  }

  cmHatch = 1;
  hatchInfo = scalloc( sizeof( s_ticfile ), 1 );
  basename = ( argv[i] != NULL ) ? argv[i++] : "";
  hatchInfo->file = sstrdup( basename );
  hatchInfo->anzdesc = 1;
  hatchInfo->desc =
      srealloc( hatchInfo->desc, ( hatchInfo->anzdesc ) * sizeof( &hatchInfo->desc ) );
  hatchInfo->desc[0] = sstrdup( "-- description missing --" );

  /*  Check filename for 8.3, warn if not */
  basename = strrchr( hatchInfo->file, PATH_DELIM );
  if( basename == NULL )
    basename = hatchInfo->file;
  else
    basename++;
  if( ( extdelim = strchr( basename, '.' ) ) == NULL )
    extdelim = basename + strlen( basename );

  if( extdelim - basename > 8 || strlen( extdelim ) > 4 )
  {
    if( !quiet )
      fprintf( stderr, "Warning: hatching file with non-8.3 name!\n" );
  }
  basename = ( argv[i] != NULL ) ? argv[i++] : "";
  hatchInfo->area = sstrdup( basename );

  if( argc - i == 0 )
  {
    return 1;
  }
  if( stricmp( argv[i], "replace" ) == 0 )
  {
    if( ( i < argc - 1 ) && ( stricmp( argv[i + 1], "desc" ) != 0 ) )
    {
      i++;
      hatchInfo->replaces = sstrdup( argv[i] );
    }
    else
    {
      basename = strrchr( hatchInfo->file, PATH_DELIM ) ?
          strrchr( hatchInfo->file, PATH_DELIM ) + 1 : hatchInfo->file;

      hatchInfo->replaces = sstrdup( basename );
    }
    if(i < argc - 1)
    {
        i++;
    }
  }
  if( stricmp( argv[i], "desc" ) != 0 )
    return 1;
  else if( argc - i < 2 )
    return 1;
  i++;
  nfree( hatchInfo->desc[0] );
  hatchInfo->desc[0] = sstrdup( argv[i] );
#ifdef __NT__
  CharToOem( hatchInfo->desc[0], hatchInfo->desc[0] );
#endif
  i++;
  if( argc - i != 0 )
  {
    hatchInfo->anzldesc = 1;
    hatchInfo->ldesc =
        srealloc( hatchInfo->ldesc, ( hatchInfo->anzldesc ) * sizeof( &hatchInfo->ldesc ) );
    hatchInfo->ldesc[0] = sstrdup( argv[i] );

#ifdef __NT__
    CharToOem( hatchInfo->ldesc[0], hatchInfo->ldesc[0] );
#endif

  }
  return 1;
}

void start_help( void )
{
  printf( "%s\n", versionStr );
  printf( "\nUsage: htick [options] [command]\n"
          "Options: -q              Quiet mode (display only urgent messages to console)\n"
          "         -v              Print program name and version to screen and exit\n"
          "         -h              Print program usage information to screen and exit\n"
          "         -c <conf-file>  Specify an alternative configuration file\n"
          "Commands:\n"
          " toss [annfecho <file>]  Do tossing. [Announce new fileechos in text file]\n"
          " toss -b                 Toss bad tics\n"
          " scan                    Scan netmail area for mails to filefix and process\n"
          "                         filefix commands contained inside the mails\n"
          " ffix [-f] [-s] <addr> <command> - Send command to our filefix from the name\n"
          "                         of <addr>\n"
          "                         -f - also send the same command to the link's filefix\n"
          "                         -s - do not send reply from our filefix to the link\n"
          " ffix! <addr> <command>  The same as 'ffix -f <addr> <command>'\n"
          " relink <pattern> <addr> Refresh subscription for areas matching the pattern\n"
          " resubscribe <pattern> <fromaddr> <toaddr> - move subscription from one link\n"
          "                         to another for the areas matching the pattern\n"
          " resubscribe -f file <fromaddr> <toaddr> - move subscription from one link\n"
          "                         to another of the areas matching the area patterns\n"
          "                         listed in this file with one pattern on a line\n"
          " clean                   Clean passthrough dir and old files in fileechos\n"
          " announce                Announce new files\n"
          " send <file> <filearea> <address> Send file from filearea to address\n"
          " hatch <file> <area> [replace [<filemask>]] [desc <desc> [<ldesc>]]\n"
          "                         Hatch file into area using description for file,\n"
          "                         if \"replace\" is present, add replace field to TIC;\n"
          "                         if <filemask> is not present then put <file> in field;\n"
          "                         if \"desc\" is present, add description(s);\n"
          "                         a short one line <desc> and a long <ldesc>.\n"
          "                         One may use the following as <desc>:\n"
          "                           @@BBS to use first line from files.bbs\n"
          "                           @@DIZ to use 1st line from the contained file_id.diz\n"
          "                           @@<file> to use first line from <file>\n"
          "                           @BBS to load from files.bbs\n"
          "                           @DIZ to load from the contained file_id.diz\n"
          "                           @<file> to load from <file>\n"
          "                         The last three items may be used as <ldesc>.\n"
          " filelist <file> [<dirlist>] Generate filelist which includes all files in base\n"
          "                             <dirlist> - list of paths to files from filelist\n" );
}

e_exitCode processCommandLine( int argc, char **argv )
{
  int i = 0;
  e_exitCode rc = ex_OK;

  if( argc == 1 )
  {
    start_help(  );
    return ex_Help;
  }

  while( i < argc - 1 )
  {
    i++;
    if( !strcmp( argv[i], "-h" ) )
    {
      start_help(  );
      return ex_Help;
    }
    if( !strcmp( argv[i], "-v" ) )
    {
      printf( "%s\n", versionStr );
      return ex_Help;
    }
    if( !strcmp( argv[i], "-q" ) )
    {
      quiet = 1;
      continue;
    }
    if( !strcmp( argv[i], "-c" ) )
    {
      i++;
      if( argv[i] != NULL )
        xstrcat( &cfgFile, argv[i] );
      else
      {
        fprintf(stderr, "Parameter missing after \"%s\"!\n", argv[i - 1]);
        rc = ex_Error;
      }
      continue;
    }
    if( stricmp( argv[i], "toss" ) == 0 )
    {
      cmToss = 1;
      if( i < argc - 1 )
      {
        if( stricmp( argv[i + 1], "-b" ) == 0 )
        {
          cmToss = 2;
          i++;
        }
      }
      continue;
    }
    else if( stricmp( argv[i], "scan" ) == 0 )
    {
      cmScan = 1;
      continue;
    }
    else if( stricmp( argv[i], "hatch" ) == 0 )
    {
      processHatchParams( i + 1, argc, argv );
      break;
    }
    else if( stricmp( argv[i], "ffix" ) == 0 )
    {
      while ( i+1 < argc && *(argv[i+1]) == '-' )
      {
          i++;
          if (stricmp(argv[i], "-f") == 0)
              cmNotifyLink = 1;
          else if (stricmp(argv[i], "-s") == 0)
              silent_mode = 1;
          else
          {
              fprintf(stderr, "Unknown afix option \"%s\"!\n", argv[i]);
              rc = ex_Error;
          }
      }
      if( i < argc - 1 )
      {
        i++;
        parseFtnAddrZS( argv[i], &afixAddr );
        if( i < argc - 1 )
        {
          i++;
          xstrcat( &afixCmd, argv[i] );
        }
        else
        {
          fprintf(stderr, "Parameter missing after \"%s\"!\n", argv[i]);
          rc = ex_Error;
        }
      }
      cmAfix = 1;
      continue;
    }
    else if( stricmp( argv[i], "ffix!" ) == 0 )
    {
      if( i < argc - 1 )
      {
        i++;
        parseFtnAddrZS( argv[i], &afixAddr );
        if( i < argc - 1 )
        {
          i++;
          xstrcat( &afixCmd, argv[i] );
        }
        else
        {
          fprintf(stderr, "Parameter missing after \"%s\"!\n", argv[i]);
          rc = ex_Error;
        }
      }
      cmNotifyLink = 1;
      cmAfix = 1;
      continue;
    }
    else if( stricmp( argv[i], "relink" ) == 0 )
    {
      if( i < argc - 1 )
      {
        i++;
        xstrcat( &relinkPattern, argv[i] );
        if( i < argc - 1 )
        {
          i++;
          parseFtnAddrZS( argv[i], &relinkFromAddr );
          cmRelink = modeRelink;
        }
        else
        {
          fprintf(stderr, "Address missing after \"%s\"!\n", argv[i]);
          rc = ex_Error;
        }
      }
      else
      {
        fprintf(stderr, "Pattern missing after \"%s\"!\n", argv[i]);
        rc = ex_Error;
      }
      continue;
    }
    else if( stricmp( argv[i], "resubscribe" ) == 0 )
    {
      if (i < argc-1)
      {
        i++;
        if (stricmp(argv[i], "-f") == 0)
        {
          if (i < argc-1)
          {
            i++;
            if (fexist(argv[i]))
            {
              if (fsize(argv[i]) == 0)
              {
                fprintf(stderr, "File \"%s\" is empty\n", argv[i]);
                rc = ex_Error;
              }
              xstrcat(&resubscribePatternFile, argv[i]);
              cmRelink = modeResubsribeWithFile;
            }
            else
            {
              fprintf(stderr, "File \"%s\" does not exist\n", argv[i]);
              rc = ex_Error;
            }
          }
          else
          {
            fprintf(stderr, "Path missing after -f\n");
            rc = ex_Error;
          }
        }
        else
        {
          xstrcat(&relinkPattern, argv[i]);
          cmRelink = modeResubsribeWithPattern;
        }
        if (i < argc-1)
        {
          i++;
          parseFtnAddrZS(argv[i], &relinkFromAddr);
          if (i < argc-1)
          {
            i++;
            parseFtnAddrZS(argv[i], &relinkToAddr);
          }
          else
          {
            fprintf(stderr, "Address missing after \"%s\"!\n", argv[i]);
            rc = ex_Error;
          }
        }
        else
        {
          fprintf(stderr, "Address missing after \"%s\"!\n", argv[i]);
          rc = ex_Error;
        }
      }
      else
      {
        fprintf(stderr, "Pattern missing after \"%s\"!\n", argv[i]);
        rc = ex_Error;
      }
    }
    else if( stricmp( argv[i], "send" ) == 0 )
    {
      cmSend = 1;
      i++;
      strcpy( sendfile, ( argv[i] != NULL ) ? argv[i++] : "" );
      strcpy( sendarea, ( argv[i] != NULL ) ? argv[i++] : "" );
      strcpy( sendaddr, ( argv[i] != NULL ) ? argv[i] : "" );
      continue;
    }
    else if( stricmp( argv[i], "filelist" ) == 0 )
    {
      cmFlist = 1;
      if( i < argc - 1 )
      {
        i++;
        flistfile = sstrdup( argv[i] );
        if( i < argc - 1 )
        {
          i++;
          dlistfile = sstrdup( argv[i] );
        }
      }
      continue;
    }
    else if( stricmp( argv[i], "announce" ) == 0 )
    {
      cmAnnounce = 1;
      continue;
    }
    else if( stricmp( argv[i], "annfecho" ) == 0 )
    {
      if( i < argc - 1 )
      {
        i++;
        cmAnnNewFileecho = 1;
        strcpy( announcenewfileecho, argv[i] );
      }
      else
      {
        fprintf(stderr, "Insufficient number of arguments\n");
        rc = ex_Error;
      }
      continue;
    }
    else if( stricmp( argv[i], "clean" ) == 0 )
    {
      cmClean = 1;
      continue;
    }
    else
    {
      fprintf(stderr, "Unrecognized command line option %s!\n", argv[i]);
      rc = ex_Error;
    }

  } /* endwhile */
  return rc;
}

void processConfig(  )
{
  char *buff = NULL;
  int exitflag = 0;

  setvar( "module", "htick" );
  SetAppModule( M_HTICK );
  config = readConfig( cfgFile );
  if( NULL == config )
  {
    nfree( cfgFile );
    fprintf( stderr, "Config file not found\n" );
    exit( 1 );
  };


  /* open Logfile */

  initLog( config->logFileDir, config->logEchoToScreen, config->loglevels,
           config->screenloglevels );
  setLogDateFormat( config->logDateFormat );
  htick_log = openLog( LogFileName, versionStr );       /* if failed: openLog() prints a message to stderr */
  if( htick_log && quiet )
    htick_log->logEcho = 0;

  if( !sstrlen( config->inbound ) && !sstrlen( config->localInbound )
      && !sstrlen( config->protInbound ) )
  {
    w_log( LL_CRIT, "You must set inbound, protInbound or localInbound in fidoconfig first" );
    exitflag = 1;
  }
  if( !sstrlen( config->outbound ) )
  {
    w_log( LL_CRIT, "You must set outbound in fidoconfig first" );
    exitflag = 1;
  }
  if( config->addrCount == 0 )
  {
    w_log( LL_CRIT, "At least one addr must be defined" );
    exitflag = 1;
  }
  if( config->linkCount == 0 )
  {
    w_log( LL_CRIT, "At least one link must be specified" );
    exitflag = 1;
  }
  if( config->fileAreaBaseDir == NULL )
  {
    w_log( LL_CRIT, "You must set FileAreaBaseDir in fidoconfig first" );
    exitflag = 1;
  }
  if( config->passFileAreaDir == NULL )
  {
    w_log( LL_CRIT, "You must set PassFileAreaDir in fidoconfig first" );
    exitflag = 1;
  }
  if( cmAnnounce && config->announceSpool == NULL )
  {
    w_log( LL_CRIT, "You must set AnnounceSpool in fidoconfig first" );
    exitflag = 1;
  }
  if( config->MaxTicLineLength && config->MaxTicLineLength < 80 )
  {
    w_log( LL_CRIT, "Parameter MaxTicLineLength (%d) in fidoconfig must be 0 or >80\n",
           config->MaxTicLineLength );
    exitflag = 1;
  }

  if( exitflag )
  {
    exit_htick( "Error(s) is found in config file, please run tparser and analize it's output.",
                1 );
  }


  if( config->lockfile )
  {
    lock_fd = lockFile( config->lockfile, config->advisoryLock );
    if( lock_fd < 0 )
    {
      disposeConfig( config );
      exit( EX_CANTCREAT );
    }
  }

  w_log( LL_START, "Start" );

  nfree( buff );

  if( config->busyFileDir == NULL )
  {
    config->busyFileDir = ( char * )smalloc( strlen( config->outbound ) + 10 );
    strcpy( config->busyFileDir, config->outbound );
    sprintf( config->busyFileDir + strlen( config->outbound ), "busy.htk%c", PATH_DELIM );
  }
  if( config->ticOutbound == NULL )
    config->ticOutbound = sstrdup( config->passFileAreaDir );
}

int main( int argc, char **argv )
{
  struct _minf m;
  int rc = 0;
  e_exitCode res;

  versionStr = GenVersionStr( "htick", VER_MAJOR, VER_MINOR, VER_PATCH, VER_BRANCH, cvs_date );

  res = processCommandLine( argc, argv );
  if( res == ex_Error )
    return EX_USAGE;
  else if( res == ex_Help )
    return EX_OK;
  processConfig(  );

  if( !( config->netMailAreas ) )
    w_log( LL_CRIT, "Netmailarea not defined, filefix not allowed!" );

  /* init areafix */
  if( !init_htickafix(  ) )
    exit_htick( "Can't init Areafix library", 1 );

  /*  init SMAPI */
  m.req_version = 0;
  m.def_zone = config->addr[0].zone;
  if( MsgOpenApi( &m ) != 0 )
  {
    exit_htick( "MsgApiOpen Error", 1 );
  }                             /*endif */

  /*  load recoding tables */
  initCharsets(  );
  getctabs( config->intab, config->outtab );
#ifdef __NT__
  SetFileApisToOEM(  );
#endif

  checkTmpDir(  );

  if( cmScan )
    scan(  );
  if( cmToss )
    toss(  );
  if( cmHatch )
    rc = hatch(  );
  if( cmSend )
    send( sendfile, sendarea, sendaddr );
  if( cmFlist )
    filelist(  );
  if( cmClean )
    Purge(  );
  if( cmAfix && config->netMailAreas )
    ffix( afixAddr, afixCmd );
  if( cmAnnounce && config->announceSpool != NULL )
    report(  );

  if (cmRelink != modeNone)
  {
    int ret;
    FILE *f;
    char *line, *fromCmd = NULL, *toCmd = NULL, *toPrint = NULL;
    unsigned int count = 0;

    w_log(LL_START, "%s has started", cmRelink == modeRelink ? "Relinking" : "Resubscribing");
    if (cmRelink != modeResubsribeWithFile)
    {
      /* modeRelink or modeResubsribeWithPattern */
      ret = relink(cmRelink, relinkPattern, relinkFromAddr, relinkToAddr, &fromCmd, &toCmd, &count);
      nfree(relinkPattern);
      if (ret)
        return 1;
    }
    else
    {
      /* modeResubsribeWithFile */
      f = fopen(resubscribePatternFile, "r");
      if (f == NULL)
      {
        fprintf(stderr, "Cannot open file \"%s\"!\n", resubscribePatternFile);
        nfree(resubscribePatternFile);
        return 1;
      }
      else
      {
        while ((line = readLine(f)) != NULL)
        {
          xstrcat(&relinkPattern, line);
          ret = relink(cmRelink, relinkPattern, relinkFromAddr, relinkToAddr, &fromCmd, &toCmd, &count);
          nfree(relinkPattern);
          if (ret)
          {
            nfree(resubscribePatternFile);
            return 1;
          }
        }
        nfree(resubscribePatternFile);
      }
    }
    if (fromCmd)
    {
      if (cmRelink == modeRelink)
        sendRelinkMsg(cmRelink, relinkFromAddr, fromCmd, smodeSubscribe);
      else
        sendRelinkMsg(cmRelink, relinkFromAddr, fromCmd, smodeUnsubscribe);
      nfree(fromCmd);
    }
    if (toCmd)
    {
      sendRelinkMsg(cmRelink, relinkToAddr, toCmd, smodeSubscribe);
      nfree(toCmd);
    }
    xscatprintf(&toPrint, "%i ", count);
    count == 1 ? xscatprintf(&toPrint, "%s has been ", af_robot->strA) : xscatprintf(&toPrint, "%ses have been", af_robot->strA);
    w_log(LL_AREAFIX, "%s %s", toPrint, cmRelink == modeRelink ? "relinked" : "resubscribed");
    nfree(toPrint);
  }

  nfree( afixCmd );
  nfree( flistfile );
  nfree( dlistfile );
  if( hatchInfo )
  {
    disposeTic( hatchInfo );
    nfree( hatchInfo );
  }
  /*  deinit SMAPI */
  MsgCloseApi(  );

  doneCharsets(  );
  w_log( LL_STOP, "End" );
  closeLog(  );
  if( config->lockfile )
  {
    FreelockFile( config->lockfile, lock_fd );
  }
  disposeConfig( config );
  nfree( versionStr );

  return rc;
}
